cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -134,6 +142,55 @@
     this.tasks = new LinkedList<>();
     this.metricsCollectionService = metricsCollectionService;
     this.programCompletionNotifiers = Collections.emptySet();
+    this.cConf = cConf;
+    this.txRunner = transactionRunner;
+    this.store = store;
+  }
+
+  @Override
+  protected void doStartUp() throws Exception {
+    super.doStartUp();
+    AtomicBoolean completed = new AtomicBoolean();
+    AtomicReference<AppMetadataStore.Cursor> cursorRef = new AtomicReference<>(AppMetadataStore.Cursor.EMPTY);
+    int limit = cConf.getInt(Constants.RuntimeMonitor.INIT_BATCH_SIZE);
+    long startTs = System.currentTimeMillis();
+
+    try {
+      while (!completed.get()) {
+        TransactionRunners.run(txRunner, context -> {
+          AppMetadataStore store = AppMetadataStore.create(context);
+          completed.set(true);
+          store.scanActiveRuns(cursorRef.get(), limit, (cursor, runRecordDetail) -> {
+            if (runRecordDetail.getStartTs() > startTs) {
+              return;
+            }
+            completed.set(false);
+            try {
+              if (runRecordDetail.getStatus() == ProgramRunStatus.STARTING) {
+                // It is unknown what is the state of program runs in STARTING state.
+                // A STARTING message is published again to retry STARTING logic.
+                ProgramOptions programOptions =
+                  new SimpleProgramOptions(runRecordDetail.getProgramRunId().getParent(),
+                                           new BasicArguments(runRecordDetail.getSystemArgs()),
+                                           new BasicArguments(runRecordDetail.getUserArgs()));
+                programStateWriter.start(runRecordDetail.getProgramRunId(),
+                                         programOptions,
+                                         null,
+                                         this.store.loadProgram(runRecordDetail.getProgramRunId().getParent()));
+                LOG.debug("Retrying to run {}.", runRecordDetail.getProgramRunId());
+              }
+            } catch (Exception e) {
+              ProgramRunId programRunId = runRecordDetail.getProgramRunId();
+              LOG.warn("Retry failed for run {}. Marking it as failed.", programRunId, e);
+              programStateWriter.error(programRunId, e);
+            }
+            cursorRef.set(cursor);
+          });
+        });
+      }
+    } catch (Exception e) {
+      LOG.error("Retry failed.", e);
```

Comment:
when will this happen ? Shall we make the error message more detailed ?

Reply:
Done!